src/drivers/intel/dw-uart.c

```diff
@@ -160,3 +175,113 @@ void dw_uart_write_word(uint32_t word)
 		uart_write(dev, SUE_UART_REG_THR, outchar);
 	}
 }
+
+#define DW_UART_FIFO_SIZE 64
+#define DW_UART_FIFO_THRESHOLD (DW_UART_FIFO_SIZE / 2)
+/* Actually the FIFO size can be read out */
+
+#define DW_UART_RING_SIZE 32
+static uint8_t dw_uart_ring[DW_UART_RING_SIZE];
+/* write to head, read from tail */
+static bool dw_uart_ring_empty = true;
+static unsigned int dw_uart_ring_head;
+static unsigned int dw_uart_ring_tail;
+
+#define min(a, b) ({		\
+	typeof(a) __a = (a);	\
+	typeof(b) __b = (b);	\
+	__a > __b ? __b : __a;	\
+})
+
+static int dw_uart_write_fifo(const uint8_t *data, unsigned int size)
+{
+	struct dw_uart_device * const dev = &uart_dev;
+	unsigned int i, count = DW_UART_FIFO_SIZE -
+		uart_read(dev, SUE_UART_REG_TFL);
+
+	if (count > size)
+		count = size;
+
+	for (i = 0; i < count; i++, data++) {
+		uint32_t value = *data;
+
+		uart_write(dev, SUE_UART_REG_THR, value);
+	}
+
+	return size - count;
+}
+
+static void dw_uart_wait(void)
+{
+	struct dw_uart_device * const dev = &uart_dev;
+
+	while (uart_read(dev, SUE_UART_REG_TFL) > DW_UART_FIFO_THRESHOLD)
+		;
+}
+
+int dw_uart_write_nowait(const uint8_t *data, int size)
+{
+	unsigned int tail_room, head_room, count;
+
+	if (size <= 0 ||
+	    (!dw_uart_ring_empty &&
+	     dw_uart_ring_tail == dw_uart_ring_head))
+		/* No data or ring full */
+		return size;
+
+	if (dw_uart_ring_empty) {
+		/* Ring empty, see if we can write any data in the FIFO */
+		size = dw_uart_write_fifo(data, size);
+		if (size <= 0)
+			return size;
+	}
+
+	/*
+	 * FIFO was filled, no threshold IRQ has triggered yet, ring might have
+	 * data
+	 */
+
+	if (dw_uart_ring_tail <= dw_uart_ring_head) {
+		head_room = sizeof(dw_uart_ring) - dw_uart_ring_head;
+		tail_room = dw_uart_ring_tail;
+	} else {
+		head_room = dw_uart_ring_tail - dw_uart_ring_head;
+		tail_room = 0;
+	}
+
+	count = min(head_room, size);
+	arch_memcpy(dw_uart_ring + dw_uart_ring_head, data, count);
+
+	dw_uart_ring_head += count;
+	if (dw_uart_ring_head == sizeof(dw_uart_ring))
+		dw_uart_ring_head = 0;
+
+	size -= count;
+
+	if (size && tail_room) {
+		count = min(tail_room, size);
+		arch_memcpy(dw_uart_ring, data + head_room, count);
+		dw_uart_ring_head = count;
+		size -= count;
+	}
+
+	dw_uart_ring_empty = false;
+
+	return size;
+}
+
+/* Block until all the data is at least in the ring buffer */
+int dw_uart_write(const uint8_t *data, int size)
```

Comment:
so for baseFW users we can pass in a dev * here. See ssp.c and dw-dma.c (they also have probe() and remove()).